File: Python/Archive/2. Add Two Numbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        return self.helper(l1, l2, 0)
    
    def helper(self, l1, l2, c):
        res = None
        if l1 or l2 or c:
            l1 = ListNode(0) if not l1 else l1
            l2 = ListNode(0) if not l2 else l2
            total = l1.val + l2.val + c
            res = ListNode(total%10)
            res.next = self.helper(l1.next, l2.next, total//10)
        return res

File: C++/Linked List/Add Two Numbers Series/2. Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        return helper(l1, l2, 0);
    }
    
    ListNode* helper(ListNode* l1, ListNode* l2, int c) {
        ListNode* res = nullptr;
        if (l1 || l2 || c) {
            l1 = l1 ? l1 : new ListNode(0);
            l2 = l2 ? l2 : new ListNode(0);
            int total = l1->val + l2->val + c;
            res = new ListNode(total%10);
            res->next = helper(l1->next, l2->next, total/10);
        }
        return res;
    }
};